src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,284 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FindSymbols
+{
+    internal partial class FindReferencesSearchEngine
+    {
+        /// <summary>
+        /// Represents the set of symbols that the engine is searching for.  While the find-refs engine is passed an
+        /// initial symbol to find results for, the engine will often have to 'cascade' that symbol to many more symbols
+        /// that clients will also need.  This includes:
+        /// <list type="number">
+        /// <item>Cascading to all linked symbols for the requested symbol.  This ensures a unified set of results for a
+        /// particular symbol, regardless of what project context it was originally found in.</item>
+        /// <item>Symbol specific cascading.  For example, when searching for a named type, reference to that named type
+        /// will be found through its constructors.</item>
+        /// <item>Cascading up and down the inheritance hierarchy for members (e.g. methods, properties, events). This
+        /// is controllable through the <see cref="FindReferencesSearchOptions.UnidirectionalHierarchyCascade"/>
+        /// option.</item>
+        /// </list>
+        /// </summary>
+        private abstract class SymbolSet
```

Comment:
this is about 50% of the PR.  It is the abstraction that allows us to represent the set of symbols we're searching for, and which allows us to update it incrementally as we process projects.  The other 50% is in the engine itself and how the core loop is updated to use this.